sematic/scheduling/job_details.py

```diff
@@ -233,3 +262,289 @@ class JobDetails:
+            return self.get_active_status_from_pods(
+                latest_summary=latest_summary,
+                current_pods=self.current_pods,
+                most_recent_condition=most_recent_condition,
+                previous_pod_name=self.previous_pod_name,
+                previous_node_name=self.previous_node_name,
+                last_updated_epoch_seconds=last_updated_epoch_seconds,
+            )
+
+    @classmethod
+    def _get_job_failed_status(
+        cls,
+        latest_summary: Optional[PodSummary],
+        last_updated_epoch_seconds: float,
+    ):
+        state_name = KubernetesJobState.Failed
+        description = "Job failed. "
+        pod_name = latest_summary.pod_name if latest_summary is not None else "UNKNOWN"
+
+        is_premature_0_exit = (
+            latest_summary is not None
+            and latest_summary.container_exit_code == 0
+            and latest_summary.has_infra_failure
+        )
+        if is_premature_0_exit:
+            description += (
+                "The container exited with a 0 exit code, "
+                "but Sematic did not record the terminal state from the worker. "
+                "Did the code inside the Sematic func contain a forced premature exit "
+                "(ex: sys.exit(0), os._exit(0))?"
+            )
+        elif (
+            latest_summary is not None
+            and latest_summary.container_condition_message is not None
+        ):
+            description += latest_summary.container_condition_message + "."
+
+        # Warning instead of error because it's normal for user jobs to fail
+        # even when Sematic/the server is healthy.
+        logger.warning("Worker pod %s failed: %s", pod_name, description)
+
+        return JobStatus(
+            state=state_name,
+            message=description,
+            last_updated_epoch_seconds=last_updated_epoch_seconds,
+        )
+
+    @classmethod
+    def get_active_status_from_pods(
+        cls,
+        latest_summary: PodSummary,
+        current_pods: List[PodSummary],
+        most_recent_condition: Optional[str],
+        previous_pod_name: Optional[str],
+        previous_node_name: Optional[str],
+        last_updated_epoch_seconds: float,
+    ) -> JobStatus:
+        state_name = KubernetesJobState.Running
+        if latest_summary.node_name is None:
+            description = f"Job is running with pod {latest_summary.pod_name}."
+        else:
+            description = (
+                f"Job is running with pod {latest_summary.pod_name} on "
+                f"node {latest_summary.node_name}."
+            )
+
+        if (
+            previous_node_name is not None
+            and previous_node_name != latest_summary.node_name
+        ):
+            description += (
+                " The pod was restarted by Kubernetes without Sematic's "
+                "intervention, and was assigned to a new Kubernetes node. "
+                "This can have many causes."
```

Comment:
Add a link to a resource listing these causes, or a link to Discord to report this?

Reply:
I don't know an exhaustive list of causes, but one of them is total node failure. I worry that if we list that as the only one, people will assume that's what it *has* to mean, though k8s protocols reserve the right for it to happen for whatever reason k8s determines it necessary.